### PR TITLE
[ci full] fix various nimbus gradle plugin issues

### LINKED ISSUE #5341: Nimbus gradle plugin CI job failing
See https://github.com/mozilla/application-services/runs/10729196986

This shouldn't impact our releases until AC stops using their own nimbus gradle plugin.

this ticket is also so that it's clear for readers why our CI on main is red

cc @jeddai

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/SYNC-3566)

### PR DESCRIPTION
**Pull request by jeddai**
Fixes #5341, [EXP-3054](https://mozilla-hub.atlassian.net/browse/EXP-3054), [EXP-3055](https://mozilla-hub.atlassian.net/browse/EXP-3055), [EXP-3056](https://mozilla-hub.atlassian.net/browse/EXP-3056), [EXP-3065](https://mozilla-hub.atlassian.net/browse/EXP-3065)

### Pull Request checklist ###

- [x] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes a changelog entry in [CHANGES_UNRELEASED.md](../CHANGES_UNRELEASED.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [x] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[ff-android: firefox-android-branch-name]` and/or `[fenix: fenix-branch-name]` to the PR title.

### CODE REVIEW

File: tools/nimbus-gradle-plugin/build.gradle [hunk 1]
```diff
@@ -136,3 +143,10 @@ publishing {
         }
     }
 }
+
+publishing.publications.withType(MavenPublication).each {publication ->
+    def checkFileSizeTask = task "checkLibSizeForMavenArtifact-${publication.artifactId}"(type: Exec) {
+        commandLine "${rootProject.projectDir}/automation/check_artifact_size.sh", project.buildDir, publication.artifactId
+    }
+    checkMavenArtifacts.dependsOn(checkFileSizeTask)
+}
```
Comment (tarikeshaq) [inline tools/nimbus-gradle-plugin/build.gradle:152]: nano-nit: missing new line at the end of file

File: tools/nimbus-gradle-plugin/src/main/groovy/org/mozilla/appservices/tooling/nimbus/NimbusGradlePlugin.groovy [hunk 2]
```diff
@@ -213,13 +199,11 @@ class NimbusPlugin implements Plugin<Project> {
...
             def successfulHost = tryDownload(archive.getParentFile(), archive.getName(),
-                    // …the latest one from github.
-                    "https://github.com/mozilla/application-services/releases/download/v$asVersion",
-                    "https://github.com/mozilla/application-services/releases/latest/download"
+                    "https://github.com/mozilla/application-services/releases/download/v$asVersion"
             )
 
             if (successfulHost == null) {
-                throw java.io.IOException("Unable to download nimbus-fml tooling")
+                throw java.io.IOException("Unable to download nimbus-fml tooling. If you are using a development version of the Nimbus Gradle Plugin, please set applicationServicesDir appropriately.")
```
Comment (tarikeshaq) [inline tools/nimbus-gradle-plugin/src/main/groovy/org/mozilla/appservices/tooling/nimbus/NimbusGradlePlugin.groovy:206]: Could we maybe describe what the setting of `applicationServicesDir` looks like? is this the same as autopublish flow in fenix, where we set property in the `local.properties` (i.e https://github.com/mozilla-mobile/fenix#auto-publication-workflow-for-android-components-and-application-services)

if so maybe just adding `please set applicationServicesDir in the local.properties appropriately` could help developers quickly understand it's the same thing their used to do

**Review by tarikeshaq — COMMENTED** [review]
LGTM from a CI is passing perspective 🎉 I'll leave approval for @jhugman re: the groovy/gradle changes

File: tools/nimbus-gradle-plugin/src/main/groovy/org/mozilla/appservices/tooling/nimbus/NimbusGradlePlugin.groovy (hunk 2 above)
Comment (jeddai) [reply tools/nimbus-gradle-plugin/src/main/groovy/org/mozilla/appservices/tooling/nimbus/NimbusGradlePlugin.groovy:206]: Actually it doesn't need to go into local.properties, necessarily, it'll need to be in the nimbus block. I've got a little snippet that helps with it and plan to include it when we move to upgrade a-c/focus/fenix to use this.

Comment (tarikeshaq) [reply tools/nimbus-gradle-plugin/src/main/groovy/org/mozilla/appservices/tooling/nimbus/NimbusGradlePlugin.groovy:206]: Awesome!! If the snippet is easy to reference on its own, is it worth putting the snippet in the changelog?

Comment (jeddai) [reply tools/nimbus-gradle-plugin/src/main/groovy/org/mozilla/appservices/tooling/nimbus/NimbusGradlePlugin.groovy:206]: Well it's not relevant in all cases, specifically not relevant for a-s modules that might be using this

Comment (tarikeshaq) [reply tools/nimbus-gradle-plugin/src/main/groovy/org/mozilla/appservices/tooling/nimbus/NimbusGradlePlugin.groovy]: Love the new error!

Comment (jeddai) [reply tools/nimbus-gradle-plugin/src/main/groovy/org/mozilla/appservices/tooling/nimbus/NimbusGradlePlugin.groovy]: Extreme specificity wins again!

**Review by jhugman — COMMENTED** [review]
This is ok: I've left a couple of comments.

Also: does this do anything to fix [EXP-3058](https://mozilla-hub.atlassian.net/browse/EXP-3058)?

**Comment by jhugman** [comment]
Testing with patches from https://github.com/mozilla/application-services/pull/5173#discussion_r1068180582

**Comment by jeddai** [comment]
> Also: does this do anything to fix [EXP-3058](https://mozilla-hub.atlassian.net/browse/EXP-3058)?

So sorry, forgot to add that into the PR. In short, the sourceSets has to have the generated folder added, and the project has to depend on the nimbus project. The `R` class is still an issue though (at least for a-s, and I'd assume a-c as well)

[update_tabs_component_to_use_nimbus.patch](https://github.com/mozilla/application-services/files/10520450/update_tabs_component_to_use_nimbus.patch)

File: tools/nimbus-gradle-plugin/src/main/groovy/org/mozilla/appservices/tooling/nimbus/NimbusGradlePlugin.groovy [hunk 3]
```diff
@@ -116,7 +102,7 @@ abstract class NimbusPluginExtension {
         var cacheDir = this.cacheDir.getOrNull()
         if (cacheDir == null)
             return null
-        return [project.rootDir, cacheDir].join(File.separator)
+        return [project.projectDir, cacheDir].join(File.separator)
```
[the remark this answers is not in the document]
Comment (jeddai) [reply tools/nimbus-gradle-plugin/src/main/groovy/org/mozilla/appservices/tooling/nimbus/NimbusGradlePlugin.groovy:105]: Hm that's a good point. I'd agree it should probably be the root directory, but yeah setting up a cache directory by default definitely couldn't hurt. I'll add that

**Comment by jeddai** [comment]
I added the default cache directory, please let me know if you have any more questions @jhugman!

**Review by jhugman — APPROVED** [review]
Sorry, I thought I'd approved this already.

I'm not sure I understand about [EXP-3058](https://mozilla-hub.atlassian.net/browse/EXP-3058), but that shouldn't hold up this PR.

Good job, Charlie. Thank you.

**Comment by jeddai** [comment]
I'll remove 3058 and we can discuss it more later!

**Pull request merged**